runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/SplittableDoFnOperator.java

```diff
@@ -141,21 +212,182 @@ public void outputWindowedValue(
+    // when we are shut down
+    continuationTimerService.registerEventTimeTimer(
+        TimerInternals.TimerData.of(
+            "continuation",
+            stateNamespace,
+            BoundedWindow.TIMESTAMP_MAX_VALUE,
+            TimeDomain.EVENT_TIME),
+        BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+  }
+
+  /**
+   * Process a restriction. Returns true if the restriction is exhausted, false otherwise.
+   */
+  private boolean processRestriction(
+      ValueState<WindowedValue<InputT>> elementState,
+      ValueState<RestrictionT> restrictionState,
+      WatermarkHoldState holdState,
+      KV<WindowedValue<InputT>, RestrictionT> elementAndRestriction) {
+    final TrackerT tracker = invoker.invokeNewTracker(elementAndRestriction.getValue());
+    SplittableProcessElementInvoker<InputT, OutputT, RestrictionT, TrackerT>.Result result =
+        processElementInvoker.invokeProcessElement(
+            invoker, elementAndRestriction.getKey(), tracker);
+
+    // Save state for resuming.
+    if (result.getResidualRestriction() == null) {
+      // All work for this element/restriction is completed. Clear state and release hold.
+      elementState.clear();
+      restrictionState.clear();
+      holdState.clear();
+      return true;
+    }
+
+    restrictionState.write(result.getResidualRestriction());
+    Instant futureOutputWatermark = result.getFutureOutputWatermark();
+    if (futureOutputWatermark == null) {
+      futureOutputWatermark = elementAndRestriction.getKey().getTimestamp();
+    }
+    holdState.add(futureOutputWatermark);
+
+    return false;
+  }
+
+  @Override
+  public void onEventTime(
+      InternalTimer<Object, TimerInternals.TimerData> flinkTimer) throws Exception {
+    if (flinkTimer.getTimestamp() != BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
+      // it's not our "last-resort" timer
+      return;
+    }
+
+    TimerInternals.TimerData timer = flinkTimer.getNamespace();
+    StateNamespace stateNamespace = timer.getNamespace();
+
+    ValueState<WindowedValue<InputT>> elementState =
+        stateInternals.state(stateNamespace, elementTag);
+    ValueState<RestrictionT> restrictionState =
+        stateInternals.state(stateNamespace, restrictionTag);
+    WatermarkHoldState holdState = stateInternals.state(stateNamespace, watermarkHoldTag);
+
+    boolean restrictionExhausted;
+    do {
+      // This is not the first ProcessElement call for this element/restriction - rather,
+      // this is a timer firing, so we need to fetch the element and restriction from state.
+      elementState.readLater();
+      restrictionState.readLater();
+
+      KV<WindowedValue<InputT>, RestrictionT> elementAndRestriction =
+          KV.of(elementState.read(), restrictionState.read());
+
+      restrictionExhausted =
+          processRestriction(elementState, restrictionState, holdState, elementAndRestriction);
+
+    } while (!restrictionExhausted);
```

Comment:
Hm, this sounds like a serious issue, because SDF is specifically intended to support infinite outputs per element. Do you think it's feasible to fix Flink to support this (e.g. optionally block shutdown on outstanding timers for an operator)?